MinPlatformPkg/Test/Library/TestPointCheckLib/StandaloneMmTestPointCheckLib.c

```diff
@@ -0,0 +1,429 @@
+  IN OUT UINTN   *CommBufferSize  OPTIONAL
+  )
+{
+  TEST_POINT_SMM_COMMUNICATION_HEADER  CommData;
+  UINTN                                TempCommBufferSize;
+
+  //
+  // If input is invalid, stop processing this SMI
+  //
+  if ((CommBuffer == NULL) || (CommBufferSize == NULL)) {
+    return EFI_SUCCESS;
+  }
+
+  TempCommBufferSize = *CommBufferSize;
+
+  if (TempCommBufferSize < sizeof (TEST_POINT_SMM_COMMUNICATION_HEADER)) {
+    DEBUG ((DEBUG_ERROR, "TestPointStandaloneMmHandler: MM communication buffer size invalid!\n"));
+    return EFI_SUCCESS;
+  }
+
+  CopyMem (&CommData, CommBuffer, sizeof (CommData));
+  if (CommData.Version != TEST_POINT_SMM_COMMUNICATION_VERSION) {
+    DEBUG ((DEBUG_ERROR, "TestPointStandaloneMmHandler: MM communication Version invalid!\n"));
+    return EFI_SUCCESS;
+  }
+
+  switch (CommData.FuncId) {
+    case TEST_POINT_SMM_COMMUNICATION_FUNC_ID_UEFI_GCD_MAP_INFO:
+      return TestPointStandaloneMmReadyToBootStandaloneMmPageProtectionHandler (CommBuffer, CommBufferSize);
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  This service verifies the system state within SMM after Exit Boot Services is invoked.
+
+  @retval EFI_SUCCESS         The test point check was performed successfully.
+**/
+EFI_STATUS
+EFIAPI
+TestPointStandaloneMmExitBootServices (
+  VOID
+  )
+{
+  return TestPointMmExitBootServices ();
+}
+
+/**
+  Register Mm Test Point handler.
+**/
+VOID
+RegisterStandaloneMmTestPointHandler (
+  VOID
+  )
+{
+  EFI_STATUS  Status;
+  EFI_HANDLE  DispatchHandle;
+
+  Status = gMmst->MmiHandlerRegister (
+                    TestPointStandaloneMmHandler,
+                    &mTestPointSmmCommunciationGuid,
+                    &DispatchHandle
+                    );
+  ASSERT_EFI_ERROR (Status);
+}
+
+/**
+  Initialize feature data.
+
+  @param[in]  Role    The test point role being requested.
+**/
+VOID
+InitData (
+  IN UINT32  Role
+  )
+{
+  EFI_STATUS  Status;
+
+  ASSERT (PcdGetSize (PcdTestPointIbvPlatformFeature) == sizeof (mFeatureImplemented));
+  CopyMem (mFeatureImplemented, PcdGetPtr (PcdTestPointIbvPlatformFeature), sizeof (mFeatureImplemented));
+
+  mTestPointStruct.Role = Role;
+  CopyMem (mTestPointStruct.FeaturesImplemented, mFeatureImplemented, sizeof (mFeatureImplemented));
+  Status = TestPointLibSetTable (
+             &mTestPointStruct,
+             sizeof (mTestPointStruct)
+             );
+  if (EFI_ERROR (Status)) {
+    if (Status != EFI_ALREADY_STARTED) {
+      ASSERT_EFI_ERROR (Status);
+    }
+  }
+}
+
+/**
+  The library constructuor.
+
+  The function does the necessary initialization work for this library
+  instance.
+
+  @retval     EFI_SUCCESS       The function always return EFI_SUCCESS.
+**/
+EFI_STATUS
+EFIAPI
+StandaloneMmTestPointCheckLibConstructor (
+  IN EFI_HANDLE           ImageHandle,
+  IN EFI_MM_SYSTEM_TABLE  *SystemTable
+  )
+{
+  InitData (PLATFORM_TEST_POINT_ROLE_PLATFORM_IBV);
+
+  RegisterStandaloneMmTestPointHandler ();
```

Comment:
I _think_ the content can be centralized? Please let me know if that is overkilling.

Reply:
Updated.